### PR TITLE
🐛Source-mssql: added support for missed data types

### LINKED ISSUE #7728: Support all types in SQL Server source
Double check that we support and have unit tests for every data types for SQL Server 15 (2019):

https://docs.microsoft.com/en-us/sql/t-sql/data-types/data-types-transact-sql?view=sql-server-ver15

https://docs.microsoft.com/en-us/sql/connect/jdbc/using-basic-data-types?view=sql-server-ver15

Might be useful to check similar PR for MySQL and Postgres
https://github.com/airbytehq/airbyte/pull/7970
https://github.com/airbytehq/airbyte/pull/8726

### PR DESCRIPTION
**Pull request by etsybaev**
## What
We had lot's of types that were not handled properly by the Source-MsSql connector.

## How
Updated to connector to properly process data type for a regular connection and for some of CDC connections.

The most of type are handled using different kind of native approaches.
But there is no support for the "hierarchyid" type even in the native SQL Server jdbc driver in DataTypes.
The DB stores it in it's own representation converted to HEX format.

Selection_306

The only way to get it as a text is to query it using internal function like this: "test_column.ToString() as test_column". 
So in the final solution we make a pre-request  for fetch a first row to get actual types and then wrap required fields with toString() function in eventual request: "test_column" -> "test_column.ToString() as test_column".

Also tested using UI for both full refresh and incremental sync types:
Full refresh:
Selection_346
Selection_347
)
Selection_349

Incremental
Selection_351
:
Selection_352
Selection_353

Also checked for cases with empty tables:
Selection_354

So lots of different tests, no fails

Selection_355

## Recommended reading order
1. `MssqlSource`
2. `MssqlSourceOperations`

## 🚨 User Impact 🚨
 🚨🚨  As we not handle values properly, some of them may come represented in another format, ex. number<->string<->date\datetime\time.
So probably full refresh would be required. 

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [x] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [x] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [x] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [x] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [x] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java
Comment (alexandr-shegeda) [inline airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java:22]: what is the purpose for declaring variable that refers to the constant value?

File: airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java
Comment (alexandr-shegeda) [inline airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java:36]: can we use Set of date-related fields instead?

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDatabase.java
Comment (alexandr-shegeda) [inline airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDatabase.java:25]: it looks like nothing really was changed here, could you possibly remove this class from the PR?

**Comment by alexandr-shegeda** [comment]
kindly reminder  that we also should bump the version for the source-mssql-strict-encrypt connector

File: airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDatabase.java
Comment (etsybaev) [reply airbyte-integrations/connectors/destination-snowflake/src/main/java/io/airbyte/integrations/destination/snowflake/SnowflakeDatabase.java:25]: this was changed automatically by fix the code-style job. I will try to remove it, but not sure if the job on CI would pass

File: airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java
Comment (etsybaev) [reply airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java:22]: > kindly reminder that we also should bump the version for the source-mssql-strict-encrypt connector

Sure, will do that along with changelog updating after got all approvals right before merging

**Comment by etsybaev** [comment]
> kindly reminder that we also should bump the version for the source-mssql-strict-encrypt connector

Sure, will do that along with changelog updating after got all approvals right before merging

File: airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java
Comment (etsybaev) [reply airbyte-db/lib/src/main/java/io/airbyte/db/DataTypeUtils.java]: There was another value at the beginning, just automatically replaced it at some moment.
Good catch, updated, thanks!

File: airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java
Comment (etsybaev) [reply airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java]: There is no way to get a native Mssql enum but updated it to use some array. It should look better now.
Thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (alexandertsukanov) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:54]: Could we move this part of code to private method?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (alexandertsukanov) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:23]: Please, add some java doc. To clarify why we need to override `setJsonField(final ResultSet resultSet, final int colIndex, final ObjectNode json)` for MSSQL Source

**Comment by alexandertsukanov** [comment]
@etsybaev  Left minor comments.

**Comment by alexandertsukanov** [comment]
LGTM

**Review by alexandertsukanov — APPROVED** [review]

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java]: done, thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:28]: done, thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (tuliren) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:140]: The name of this method is a bit confusing. What about something like `getWrappedColumn`?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (tuliren) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:145]: Maybe `hierarchyIdColumns`?

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (tuliren) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:131]: Some punctuations are missing, and making this comment less easy to understand.

```suggestion
   * There is no support for hierarchyid even in the native SQL Server JDBC driver. Its value can
   * be converted to a nvarchar(4000) data type by calling the ToString() method. So we make
   * a separate query to get hierarchyid columns, and wrap required fields with the ToString() function.
   * Reference:
   * https://docs.microsoft.com/en-us/sql/t-sql/data-types/hierarchyid-data-type-method-reference?view=sql-server-ver15#data-type-conversion
```

**Comment by tuliren** [comment]
Could you also run `./gradle format`? I think some of the code will be reformatted once you run spotless.

File: airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java
Comment (etsybaev) [reply airbyte-integrations/bases/debezium/src/main/java/io/airbyte/integrations/debezium/internals/MSSQLConverter.java]: Updated, thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:33]: Updated, thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:146]: Good catch! Updated and added test for this case, thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java:25]: Once I enabled the testCatalog = true as you'd suggested - it started failing and I was able to find a missed logic. Thanks!

File: airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/MssqlSourceDatatypeTest.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/test-integration/java/io/airbyte/integrations/source/mssql/MssqlSourceDatatypeTest.java:176]: Once I enabled the testCatalog = true as you'd suggested - it started failing and I was able to find a missed logic. Thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Unfortunately, we can not use a generic method for List as in this case we would get it enquoted like this `"test_column.ToString() AS test_column"`, 
but need to have it like this:
`test_column.ToString() AS "test_column"`.

So we need to have a custom enquoting for such case.

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Updated, thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Updated, thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Updated, thanks

**Comment by etsybaev** [comment]
> Could you also run `./gradle format`? I think some of the code will be reformatted once you run spotless.

Sure, I always run it before pushing my changes. Thanks

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (tuliren) [inline airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java:109]: Is that just using `newColumnNames` will be sufficient here and the `getWrappedColumn` call is not necessary?

It is the case in `queryTableFullRefresh` above:

```java
  List<String> newIdentifiersList = getWrappedColumn(database,
        columnNames,
        schemaName, tableName, "\"");
    String preparedSqlQuery = String
        .format("SELECT %s FROM %s", enquoteIdentifierList(newIdentifiersList),
            getFullTableName(schemaName, tableName));
```

which generates the correct columns for the select clause: `SELECT "id",test_column.ToString() as "test_column" FROM "dbo"."dbo_1_hierarchyid"` because the first `getWrappedColumn` method (with 5 parameters) already adds quotation marks to the column names.

Basically I still think the second `getWrappedColumn` (with two parameters) is unnecessary.

To answer this comment:
> Unfortunately, we can not use a generic method for List as in this case we would get it enquoted like this
> "test_column.ToString() AS test_column", but need to have it like this:
> test_column.ToString() AS "test_column".
>
> So we need to have a custom enquoting for such case.

If we compare the two enquoting methods:

```java
public String getWrappedColumn(final Connection connection, final List<String> identifiers) throws SQLException {
  final StringJoiner joiner = new StringJoiner(",");
  for (final String col : identifiers) {
    if (col.contains("()")) {
      joiner.add(col);
    } else {
      final String s = sourceOperations.enquoteIdentifier(connection, col);
      joiner.add(s);
    }

  }
  return joiner.toString();
}

protected String enquoteIdentifierList(final List<String> identifiers) {
  final StringJoiner joiner = new StringJoiner(",");
  for (final String identifier : identifiers) {
    if (identifier.contains("()")) {
      joiner.add(identifier);
    } else {
      joiner.add(getIdentifierWithQuoting(identifier));
    }
  }
  return joiner.toString();
}
```

They are not doing anything differently. Specifically the identifiers won't be get enquoted like `"test_column.ToString() AS test_column"` because 1) this column will already be partially enquoted as `test_column.ToString() AS "test_column"`, and 2) any column with the `ToString()` method will be skipped and just get added to the `joiner`.

**Review by tuliren — APPROVED** [review]

**Comment by etsybaev** [comment]
-/test-performance connector=connectors/source-mssql

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSourceOperations.java]: Updated, thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Updated, thanks!

File: airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java
Comment (etsybaev) [reply airbyte-integrations/connectors/source-mssql/src/main/java/io/airbyte/integrations/source/mssql/MssqlSource.java]: Ohhh, now I got your point. Yup, you're absolutely right. I just worked at these parts (full refresh and incremental) if little bit different periods of time and didn't notice :) 
Updated and refactored. Thanks!

**Comment by etsybaev** [comment]
Currently, we have a known issue with build on CI, but locally all checks passed. So going to release new builds

**Pull request merged**
